src/auth/awsrequestsigner.ts

```diff
@@ -0,0 +1,305 @@
+ * signing process.
+ * https://docs.aws.amazon.com/general/latest/gr/signature-version-4.html
+ */
+export class AwsRequestSigner {
+  private readonly crypto: Crypto;
+
+  /**
+   * Instantiates an AWS API request signer used to send authenticated signed
+   * requests to AWS APIs based on the AWS Signature Version 4 signing process.
+   * This also provides a mechanism to generate the signed request without
+   * sending it.
+   * @param getCredentials A mechanism to retrieve AWS security credentials
+   *   when needed.
+   * @param region The AWS region to use.
+   */
+  constructor(
+    private readonly getCredentials: () => Promise<AwsSecurityCredentials>,
+    private readonly region: string
+  ) {
+    this.crypto = createCrypto();
+  }
+
+  /**
+   * Generates the signed request for the provided HTTP request for calling
+   * an AWS API. This follows the steps described at:
+   * https://docs.aws.amazon.com/general/latest/gr/sigv4_signing.html
+   * @param amzOptions The AWS request options that need to be signed.
+   * @return A promise that resolves with the GaxiosOptions containing the
+   *   signed HTTP request parameters.
+   */
+  async getRequestOptions(amzOptions: GaxiosOptions): Promise<GaxiosOptions> {
+    if (!amzOptions.url) {
+      throw new Error('"url" is required in "amzOptions"');
+    }
+    // Stringify JSON requests. This will be set in the request body of the
+    // generated signed request.
+    const requestPayloadData =
+      typeof amzOptions.data === 'object'
+        ? JSON.stringify(amzOptions.data)
+        : amzOptions.data;
+    const url = amzOptions.url;
+    const method = amzOptions.method || 'GET';
+    const requestPayload = amzOptions.body || requestPayloadData;
+    const additionalAmzHeaders = amzOptions.headers;
+    const awsSecurityCredentials = await this.getCredentials();
+    const uri = new URL(url);
+    const headerMap = await generateAuthenticationHeaderMap({
+      crypto: this.crypto,
+      host: uri.host,
+      canonicalUri: uri.pathname,
+      canonicalQuerystring: uri.search.substr(1),
+      method,
+      region: this.region,
+      securityCredentials: awsSecurityCredentials,
+      requestPayload,
+      additionalAmzHeaders,
+    });
+    // Append additional optional headers, eg. X-Amz-Target, Content-Type, etc.
+    const headers: {[key: string]: string} = Object.assign(
+      // Add x-amz-date if available.
+      headerMap.amzDate ? {'x-amz-date': headerMap.amzDate} : {},
+      {
+        Authorization: headerMap.authorizationHeader,
+        host: uri.host,
+      },
+      additionalAmzHeaders || {}
+    );
+    if (awsSecurityCredentials.token) {
+      Object.assign(headers, {
+        'x-amz-security-token': awsSecurityCredentials.token,
+      });
+    }
+    const awsSignedReq: GaxiosOptions = {
+      url,
+      method: method,
+      headers,
+    };
+
+    if (typeof requestPayload !== 'undefined') {
+      awsSignedReq.body = requestPayload;
```

Comment:
Technically, the `typeof` check is redundant, because if it's undefined, the assignment would work as well.

Reply:
True there is no side effect in assigning this without the check. The intent is not to set `undefined` in the body. Otherwise, in the test assertions, I have to match for `undefined` body fields for requests that don't have a body. I find keeping this check cleaner.

Reply:
I see, thank you for the explanation.

Reply:
You're welcome